topics/w6-live-programming.md

```diff
@@ -1,11 +1,70 @@
+- user1@example.com -> example.com
+- john@ethz.ch -> ethz.ch
+- jane@gmail.com -> gmail.com
+- food@mensacorp.org -> mensacorp.org
+    
+```python
+#Synthesized code:¨
+    splitchar = "@"
+    domain = email[1:-1][email.find(splitchar):-1] + email[-2:len(email)]
+```
+This solution, while less elegant than the first one, is still correct. However we can see that the synthesizer has some limitations. First of all, the splitchar has to be user-defined, else synthesis fails. Secondly, the solution for some reason first removes the first and last character of the email before extracting the domain, and then adds them back in. This is obviously unnecessary and is an artifact of the synthesizers enumeration based approach - which is not able to reason about the logic of the problem.
+Lastly, lets consider a program that takes a list of integers and returns the last element.
+- [1, 3, 5, 6] -> 6
+- [3, 2, 7, 5] -> 5
+
+```python
+#Synthesized code:
+    last_elem = len(input_list) + 2 // min(input_list)
+```
+What? This is absolutely not what we were looking for. (a very simple solution would be ```input_list[-1]```) When doing the math, we can see that the expression is indeed correct for our input/output pairs. However, it is not a correct solution to the problem. This shows that the synthesizer is very bad at using list indeces. Diving deeper into more examples could reveal even more limitations of the synthesizer.
+
+While SnipPy presents an innovative fusion of live programming and programming by example, its current synthesizer
+struggles to meet the practical needs of developers.
+However, the landscape of code synthesis has changed quite dramatically since the publication of this tool. 
+New large language models like GPT, BERT or Claude are able to reason about code and problems in a much more accurate way than the limited enumeration approach of SnipPy. It would be interesting to see how the interface of SnipPy could be combined with an LLM based synthesizer. 
+Tools like github copilot function in a similar way - where the user provides a prompt and the tool is able to generate code in-editor.
+
+A promising step in this direction is LEAP, a tool that integrates LLM-based code synthesis with live programming. LEAP generates multiple code completions for a given problem, allowing users to preview the options. Upon selection, LEAP displays the chosen code with corresponding projection boxes based on sample inputs, which visualize the program state. Unlike SnipPy, LEAP primarily uses projection boxes for code verification rather than problem specification. This shift aligns better with the strengths of LLMs, which can reason about prior code context and high-level problem requirements.
+
+In conclusion, SnipPy demonstrates the potential of integrating live programming with programming by example, while its limitations reveal opportunities for more advanced synthesis techniques to enhance developer tools.
```

Comment:
The current discussion on program synthesis is solid but lacks depth beyond stating that LLMs can improve synthesis accuracy.
Also, try to reflect the limitation of SnipPy within the context of live programming. Why is liveness particularly valuable in this domain? Could the constant feedback and interactivity become a distraction for programmers, especially in tasks requiring deep focus? A critical reflection on these aspects would provide a more interesting perspective.